### PR TITLE
Fix test runner issues around our stubbed windows

### PR DESCRIPTION
**Pull request by PureWeen**
### Description of Change

- I'm not sure when this [PR](https://github.com/dotnet/maui/pull/13025) will get merged so this PR pulls in all the device test runner updates from that PR which will be helpful for other PRs that might be exhibiting unexpected, weird behavior. 
- Remove the `CreateHandlerAndAddToWindow` overload that causes an `async void` path which will cause a test to keep running after xUnit thinks the test is done running

### Backport notes

These changes only modify test runner code so it's safe to backport and will make the future a better place.

### CODE REVIEW

File: src/Controls/tests/DeviceTests/ControlsHandlerTestBase.cs [hunk 1]
```diff
@@ -322,20 +339,45 @@ protected void OnUnloaded(VisualElement frameworkElement, Action action)
...
+			catch (TimeoutException)
+			{
+				if (frameworkElement.IsLoadedOnPlatform() &&
+					frameworkElement.IsLoaded)
+				{
+					taskCompletionSource.TrySetResult(true);
+				}
+				else
+					throw;
+			}
```
Comment (mandel-macaque) [inline src/Controls/tests/DeviceTests/ControlsHandlerTestBase.cs:380]: Very very similar methods, same logic, only change is the delegate, I'm wondering if we could move that to a common method and then work with it, for example:
```csharp
protected async Task AsyncWrapper (VisualElement frameworkElement, Action<VisualElement, Action> syncAction, TimeSpan? timeOut = null)
{
	TaskCompletionSource<object> taskCompletionSource = new TaskCompletionSource<object>();
	try
 	{
 		timeOut = timeOut ?? TimeSpan.FromSeconds(2);
 		syncAction(frameworkElement, () => taskCompletionSource.TrySetResult(true));
 		await taskCompletionSource.Task.WaitAsync(timeOut.Value);
 	}
    catch (TimeoutException)
 	{
 		if (frameworkElement.IsLoadedOnPlatform() &&
 			frameworkElement.IsLoaded)
 		{
 			taskCompletionSource.TrySetResult(true);
 		}
 		else
 			throw;
 	}
}
protected async Task OnUnloadedAsync(VisualElement frameworkElement, TimeSpan? timeOut = null)
	=> AsyncWrapper(frameworkElement, OnUnloaded, timeOut);

protected async Task OnLoadedAsync(VisualElement frameworkElement, TimeSpan? timeOut = null)
	=> AsyncWrapper(frameworkElement, OnLoaded, timeOut);

```
That way, if we have a bug in the async call, we know we fixed it for both cases, and we have less code to work with.

**Review by mandel-macaque — CHANGES_REQUESTED** [review]

File: src/Controls/tests/DeviceTests/ControlsHandlerTestBase.cs (hunk 1 above)
Comment (mandel-macaque) [reply src/Controls/tests/DeviceTests/ControlsHandlerTestBase.cs:380]: I am terrible at naming, so AsyncWrapper might not be the best name, but you get the idea.

Comment (mandel-macaque) [reply src/Controls/tests/DeviceTests/ControlsHandlerTestBase.cs:380]: @PureWeen After you mentioned the diff catch condition:
```csharp
protected async Task AsyncWrapper (VisualElement frameworkElement, Action<VisualElement, Action> syncAction, Func<bool> onTimeOut, TimeSpan? timeOut = null)
{
	TaskCompletionSource<object> taskCompletionSource = new TaskCompletionSource<object>();
	try
 	{
 		timeOut = timeOut ?? TimeSpan.FromSeconds(2);
 		syncAction(frameworkElement, () => taskCompletionSource.TrySetResult(true));
 		await taskCompletionSource.Task.WaitAsync(timeOut.Value);
 	}
        catch (TimeoutException)
 	{
 		if (onTimeOut (taskCompletionSource)) {
 		 	taskCompletionSource.TrySetResult(true);
 	 	}
 	 	else
 		 	throw;
 	}
}

protected async Task OnUnloadedAsync(VisualElement frameworkElement, TimeSpan? timeOut = null)
       => AsyncWrapper(frameworkElement, 
              OnUnloaded, 
              () => !frameworkElement.IsLoadedOnPlatform() && !frameworkElement.IsLoaded), 
              timeOut);

protected async Task OnLoadedAsync(VisualElement frameworkElement, TimeSpan? timeOut = null)
       => AsyncWrapper(frameworkElement, 
              OnLoaded,  
              () => frameworkElement.IsLoadedOnPlatform() && frameworkElement.IsLoaded,
              timeOut);
```
As I said, my naming is terrible :) 

And now question, why dows the tis use object and not bool?

Comment (PureWeen) [reply src/Controls/tests/DeviceTests/ControlsHandlerTestBase.cs]: Just deleted this code and consolidated into the two already existing methods. 

Personally, I'd rather just install Rx and FromEventPattern everything  :-)

File: src/Controls/tests/DeviceTests/ControlsHandlerTestBase.cs [hunk 2]
```diff
@@ -279,63 +287,63 @@ protected Task CreateHandlerAndAddToWindow<THandler>(IElement view, Func<THandle
...
+				source.TrySetResult();
 			}
+			else
+			{
+				EventHandler loaded = null;
 
-			EventHandler loaded = null;
+				loaded = (_, __) =>
+				{
+					if (loaded != null)
```
Comment (mandel-macaque) [inline src/Controls/tests/DeviceTests/ControlsHandlerTestBase.cs:304]:
```suggestion
					if (loaded is not null)
```

File: src/Controls/tests/DeviceTests/ControlsHandlerTestBase.cs [hunk 3]
```diff
@@ -150,6 +140,9 @@ protected Task CreateHandlerAndAddToWindow<THandler>(IElement view, Func<THandle
 					{
 						IView content = window.Content;
 
+						if (window is Window w && w.Navigation.ModalStack.Count > 0)
+							content = w.Navigation.ModalStack.LastOrDefault();
```
Comment (mandel-macaque) [inline src/Controls/tests/DeviceTests/ControlsHandlerTestBase.cs:144]: content is set to not nullable (no idea if we have nullability in this file) gout LastOrDefault can return null, should we worry?

File: src/Controls/tests/DeviceTests/ControlsHandlerTestBase.cs [hunk 4]
```diff
@@ -183,9 +184,16 @@ protected Task CreateHandlerAndAddToWindow<THandler>(IElement view, Func<THandle
 						else
 							throw new Exception($"I can't work with {typeof(THandler)}");
 
-						window.Deactivated();
-						window.Destroying();
 
+						bool isActivated = (window as Window)?.IsActivated ?? false;
+						bool isDestroyed = (window as Window)?.IsDestroyed ?? false;
```
Comment (mandel-macaque) [inline src/Controls/tests/DeviceTests/ControlsHandlerTestBase.cs:189]: Are this properties part of the IWindow interface?

File: src/Controls/tests/DeviceTests/Stubs/WindowHandlerStub.iOS.cs [hunk 5]
```diff
@@ -13,37 +13,61 @@ public class WindowHandlerStub : ElementHandler<IWindow, UIWindow>, IWindowHandl
...
-					var vc =
-						(_currentView.Handler as IPlatformViewHandler)
-							.ViewController;
+				var virtualView = VirtualView;
+				var view = virtualView.Content.ToPlatform(MauiContext);
+				_currentView = virtualView.Content;
+
+				var vc =
+					(_currentView.Handler as IPlatformViewHandler)
+						.ViewController;
```
Comment (mandel-macaque) [inline src/Controls/tests/DeviceTests/Stubs/WindowHandlerStub.iOS.cs:35]:
```suggestion
					(_currentView.Handler as IPlatformViewHandler)
						?.ViewController;
```
as could mean null

File: src/TestUtils/src/DeviceTests/AssertionExtensions.cs [hunk 6]
```diff
@@ -133,5 +133,48 @@ public static Task WaitForKeyboardToShow(this IView view, int timeout = 1000)
...
+#else
+			throw new NotImplementedException();
+#endif
+		}
+
+		public static IDisposable OnLoaded(this IElement element, Action action)
+		{
+#if PLATFORM
+			if (element.Handler is IPlatformViewHandler platformViewHandler &&
+				platformViewHandler.PlatformView != null)
```
Comment (mandel-macaque) [inline src/TestUtils/src/DeviceTests/AssertionExtensions.cs:156]:
```suggestion
				platformViewHandler.PlatformView is not null)
```

**Review by mandel-macaque — CHANGES_REQUESTED** [review]

File: src/Controls/tests/DeviceTests/ControlsHandlerTestBase.cs (hunk 3 above)
Comment (PureWeen) [reply src/Controls/tests/DeviceTests/ControlsHandlerTestBase.cs:144]: I'll change this to `Last`

File: src/Controls/tests/DeviceTests/ControlsHandlerTestBase.cs (hunk 4 above)
Comment (PureWeen) [reply src/Controls/tests/DeviceTests/ControlsHandlerTestBase.cs:189]: No, these properties are currently specific to the `Controls` layer (non-interface layer). At a later date we could make these properties of `IWindow` but we are going to backport this PR so don't want to make any API changes here.

File: src/Controls/tests/DeviceTests/ControlsHandlerTestBase.cs [hunk 7]
```diff
@@ -170,7 +163,15 @@ protected Task CreateHandlerAndAddToWindow<THandler>(IElement view, Func<THandle
 
 						await OnLoadedAsync(content as VisualElement);
 
-						window.Activated();
+						if (window is Window controlsWindow)
+						{
+							if (!controlsWindow.IsActivated)
```
[the remark this answers is not in the document]
Comment (PureWeen) [reply src/Controls/tests/DeviceTests/ControlsHandlerTestBase.cs:168]: I don't want the else to run 

image

**Review by mandel-macaque — APPROVED** [review]

**Pull request merged**